packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/framework.ts

```diff
@@ -0,0 +1,177 @@
+        ...event.Data,
+        ...isCompleteResult.Data
+      }
+    };
+
+    await submitCloudFormationResponse('SUCCESS', response);
+  });
+}
+
+// invoked when completion retries are exhaused.
+async function onTimeout(timeoutEvent: any) {
+  log('timeoutHandler', timeoutEvent);
+
+  const isCompleteRequest = JSON.parse(JSON.parse(timeoutEvent.Cause).errorMessage) as AWSCDKAsyncCustomResource.IsCompleteRequest;
+  await submitCloudFormationResponse('FAILED', isCompleteRequest, {
+    reason: 'Operation timed out'
+  });
+}
+
+async function invokeUserFunction(arnEnvironment: string, payload: any) {
+  const arn = getEnv(arnEnvironment);
+  log(`executing user function ${arnEnvironment} with payload`, payload);
+  const resp = await invokeFunction({ FunctionName: arn, Payload: JSON.stringify(payload) });
+
+  log('user function response:', resp, typeof(resp));
+
+  const jsonPayload = parseJsonPayload(resp.Payload);
+  if (resp.FunctionError) {
+    log('user function threw an error:', resp.FunctionError);
+    const errorMessage = jsonPayload.errorMessage || 'error';
+    const trace = jsonPayload.trace ? `\nRemote function error: ` + jsonPayload.trace.join('\n') : '';
+
+    const e = new Error(errorMessage);
+    e.stack += trace;
+    throw e;
+  }
+
+  return jsonPayload;
+}
+
+function parseJsonPayload(payload: any): any {
+  if (!payload) { return { }; }
+  const text = payload.toString();
+  try {
+    return JSON.parse(text);
+  } catch (e) {
+    log('parse error:', e);
+    return { text };
+  }
+}
+
+function validateCfnRequest(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent) {
+  const errorPrefix = `Invalid CloudFormation custom resource event (${cfnRequest.RequestType || ''}):`;
+
+  [ 'LogicalResourceId', 'RequestId', 'RequestType', 'ResponseURL', 'StackId', 'ServiceToken' ].forEach(field => {
+    if (!(cfnRequest as any)[field]) {
+      throw new Error(`${errorPrefix} ${field} is required`);
+    }
+  });
+
+  if (cfnRequest.RequestType === 'Create' && (cfnRequest as any).PhysicalResourceId) {
+    throw new Error(`${errorPrefix} PhysicalResourceId is not allowed for "Create" events`);
+  }
+
+  if ((cfnRequest.RequestType === 'Update' || cfnRequest.RequestType === 'Delete') && !cfnRequest.PhysicalResourceId) {
+    throw new Error(`${errorPrefix} PhysicalResourceId is required for "Update" and "Delete" events`);
+  }
+}
+
+function createResponseEvent(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent, onEventResult: OnEventResponse): AWSCDKAsyncCustomResource.IsCompleteRequest {
+  //
+  // validate that onEventResult always includes a PhysicalResourceId
+
+  if (!onEventResult || !onEventResult.PhysicalResourceId) {
+    throw new Error(`onEvent response must include a PhysicalResourceId for all request types`);
+  }
+
+  // if we are in DELETE and physical ID was changed, it's an error.
+  if (cfnRequest.RequestType === 'Delete' && onEventResult.PhysicalResourceId !== cfnRequest.PhysicalResourceId) {
```

Comment:
Is it not okay to return an empty `PhysicalResourceId` on a DELETE event?

Reply:
Let me do some tests. Maybe we can just omit it for `Delete` operations.

Reply:
I'll leave it like this for now. Ok. we can always relax later.